interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,367 @@
+				}
+			}
+		case "attributes", "environment":
+			err = iface.validateUDevValueMap(value)
+		default:
+			err = errors.New(`unknown key`)
+		}
+
+		if err != nil {
+			return fmt.Errorf(`custom-device "udev-tagging" invalid %q tag: %v`, key, err)
+		}
+	}
+
+	if !hasKernelTag {
+		return errors.New(`custom-device udev tagging rule missing mandatory "kernel" key`)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) Name() string {
+	return "custom-device"
+}
+
+func (iface *customDeviceInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              customDeviceSummary,
+		BaseDeclarationPlugs: customDeviceBaseDeclarationPlugs,
+		BaseDeclarationSlots: customDeviceBaseDeclarationSlots,
+		AffectsPlugOnRefresh: true,
+	}
+}
+
+func (iface *customDeviceInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if slot.Attrs == nil {
+		slot.Attrs = make(map[string]interface{})
+	}
+	// custom-device defaults to "slot" name if unspecified
+	slot.Attrs["custom-device"] = slot.Name
+
+	var devices []string
+	err := slot.Attr("devices", &devices)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for _, device := range devices {
+		if err := iface.validateDevice(device); err != nil {
+			return err
+		}
+	}
+
+	if err := iface.validatePaths(slot, "read"); err != nil {
+		return err
+	}
+	if err := iface.validatePaths(slot, "write"); err != nil {
+		return err
+	}
+
+	var udevTaggingRules []map[string]interface{}
+	err = slot.Attr("udev-tagging", &udevTaggingRules)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for _, udevTaggingRule := range udevTaggingRules {
+		if err := iface.validateUDevTaggingRule(udevTaggingRule, devices); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	customDeviceAttr, isSet := plug.Attrs["custom-device"]
+	customDevice, ok := customDeviceAttr.(string)
+	if isSet && !ok {
+		return fmt.Errorf(`custom-device "custom-device" attribute must be a string, not %v`,
+			plug.Attrs["custom-device"])
+	}
+	if customDevice == "" {
+		if plug.Attrs == nil {
+			plug.Attrs = make(map[string]interface{})
+		}
+		// custom-device defaults to "plug" name if unspecified
+		plug.Attrs["custom-device"] = plug.Name
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	snippet := &bytes.Buffer{}
+	emitRule := func(paths []string, permissions string) {
+		for _, path := range paths {
+			fmt.Fprintf(snippet, "\"%s\" %s,\n", path, permissions)
+		}
+	}
+
+	// get all attributes without validation, since that was done before;
+	// should an error occur, we'll simply not write any rule.
+
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+	emitRule(devicePaths, "rw")
```

Comment:
I wonder whether we should also support a "read-devices" attribute with the same behavior has "devices" but emitting only "r" access, I see a few interfaces that grant read-only access to some /dev/* (cc @anonymouse64). One could  have both.

Reply:
looking at the current interfaces the situation doesn't look symmetric, most allows rw for devices, so I think asymmetric naming here is fine. (naive greps gives me:  127 rw vs 20 r)

Reply:
Added `read-devices`.

Reply:
Personally I find the `devices` vs `read-devices` confusing as it is not clear to me how they differ - IMO a single `devices` would exist and each entry would be either `read` or `write` with a list of paths, ie:

```yaml
devices:
  read:
    - /dev/foo
  write:
    - /dev/bar
```

As this then mimics `system-files` etc and is clearer IMO as to what is being granted.

Reply:
I rather like @alexmurray's proposal indeed it does make things more readable